www/waterfall_view/src/module/main.module.js

```diff
@@ -635,14 +646,96 @@ var WaterfallController = (function() {
             });
         }
 
-        renderNewData() {
+        toggleTag(tag) {
+            if (!this.$scope.tags_filter.includes(tag)) {
+                this.$scope.tags_filter.push(tag);
+            } else {
+                this.$scope.tags_filter = _.remove(this.$scope.tags_filter, function(currentTag) {
+                    return currentTag != tag;
+                });
+            }
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        isTagFiltered(tag) {
+            if (this.$scope.tags_filter.includes(tag)) {
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        clearTags() {
+            this.$scope.tags_filter = this.tags_filter = [];
+            this.$location.search("tags", this.$scope.tags_filter);
+            this.renderNewData(this.$scope.tags_filter);
+        }
+
+        setTagBuilders(currentTags, builders) {
```

Comment:
setTagBuilders does not set anything makeTagBuilders?

Reply:
you mean rename it?